refs #1498 keeping variables during configuration reload

nazim9214 opened the pull request:
fixes #1498 

On configuration reload - variables with empty values/missing variables are taken from current configuration.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nlog/nlog/1769)
<!-- Reviewable:end -->

304NotModified commented:
Thanks! 

I'm doubting, shouldnt this be optional?

nazim9214 commented:
Hmm, it's good idea! What do you think, should it be per config level (i think at 'nlog' element level) or per variable (maybe 'keepOnReload' attribute)?

304NotModified commented:
I guess on NLog level will be sufficient.

304NotModified commented:
Thanks for the change! 

Could you also add an unit test for `keepVariablesOnReload=false`? Thx in advance!

nazim9214 commented:
I have added unit test for keepVariablesOnReload=false. Also i changed implementation compared to previous. Now, if keepVariablesOnReload is provided, we copy all variables from current configuration into reloaded (new variables are remain).

nazim9214 commented:
In current implementation, when keepVariablesOnReload attribute isn't specified on reloaded config, i get this value from previous configuration. Is this a good decision?

304NotModified commented:
Sounds good to me:)

304NotModified commented:
Reviewed 4 of 5 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1769)*
<!-- Sent from Reviewable.io -->

304NotModified commented:
Thanks! It has been merged and will be released in 4.4

Could you add some docs to the wiki? Please add "introduced in NLog 4.4". Thanks in advance!

nazim9214 commented:
I have added some documentation, can you check it please? (i'm not sure in my english)
https://github.com/NLog/NLog/wiki/Configuration-file/_compare/8453ae91b65e3a650b3efe7de87c6c4b5b598eff...4d385fe3b66b27aaaf05de1b5a1c2af80116f1e7

304NotModified commented:
Thanks! Changed it just a bit!
